common/transcript.ts

```diff
@@ -0,0 +1,315 @@
+import * as Constants from "config/constants";
+import langs from 'langs';
+import { parse } from 'csv-parse/sync';
+import { pipeline } from 'node:stream/promises';
+import { split, SentenceSplitterSyntax } from "sentence-splitter";
+
+import type { Readable } from 'node:stream';
+import type { WhisperXTranscript } from "common/whisperx";
+
+///////////////////
+/// Types
+///////////////////
+
+export type CategoryId = string;
+export type SegmentId = string;
+export type SpeakerId = number;
+export type Iso6393Code = string;
+export type VideoId = string;
+
+// Use array to compress size of keys in json serialization.
+export type SentenceMetadata = [
+  SegmentId,  // Id string that is unique to one transcript.
+  SpeakerId,  // Number for the speaker.
+  number,  // Start time
+  number   // End time.
+  ];
+
+export type TranscriptVersion = 1;
+
+export type TranscriptData = {
+  version : TranscriptVersion;
+  language : string;
+  sentenceMetadata : SentenceMetadata[];
+};
+
+export type SentenceTable = {
+  [id : string] : string;
+};
+
+export type MetadataBySpeaker = {
+  speaker: number;
+  sentenceMetadata : SentenceMetadata[];
+};
+
+export type SpeakerNameMap = {
+  [speaker: number] : string;
+};
+
+export type LanguageToSentenceTable = {
+  [language : Iso6393Code] : SentenceTable;
+};
+
+///////////////////
+/// Constants
+///////////////////
+
+const WHISPERX_GRANULARITY_S = 0.001;
+
+// SMALL_TS_INCREMENT_S is a very small increment in the timestamp used to synthetically
+// advance time if time timestamps are missing.
+const SMALL_TS_INCREMENT_S = WHISPERX_GRANULARITY_S / 100;
+
+export const UnknownSpeakerNum : number = 99;
+
+const TSV_ISO6393_REGEX = /\.([a-z]{3})\.tsv$/;
+
+///////////////////
+/// Classes
+///////////////////
+export abstract class StorageAccessor {
+  listFilesByPrefix(prefix: string) : Promise<string[]>;
+  getBytes(path: string) : Promise<Buffer>;
+  writeBytes(path: string, data: string) : Promise<void>;
+}
+
+export class DiarizedTranscript {
+  category: CategoryId;
+  videoId: VideoId;
+  transcriptData: TranscriptData;
+  languageToSentenceTable: LanguageToSentenceTable;
+
+  constructor(
+      category: CategoryId,
```

Comment:
Fun typescript trick. All of these can be replaced with

```
constructor(
  readonly categoryId: CategoryId,
  readonly videoId: VideoId,
  readonly transcriptData: TranscriptData,
  readonly languageToSentenceTable: LanguageToSentenceTable
) {}

Reply:
Nice! Done.